### PR TITLE
feat(quarkus): Initial Quarkus support

### LINKED ISSUE #990: support for quarkus in JVM mode

### PR DESCRIPTION
**Pull request by astefanutti**
Fixes #990.

TODO:
* [x] Versions configuration (Camel Quarkus, Quarkus)
* [x] Proper Camel Quarkus catalog integration:
  * [x] Requires apache/camel-k-runtime#167

**Release Note**
```release-note
Initial Quarkus support (JVM mode)
```

### CODE REVIEW

File: pkg/builder/runtime/quarkus.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+	// Build the project, as the quarkus-bootstrap plugin build-tree goal
+	// requires the artifact to be installed
+	mc.AddArgument("install")
+	if err := maven.Run(mc); err != nil {
+		return errors.Wrap(err, "failure while building project")
+	}
+
+	// Call the Quarkus dependencies plugin
+	mc.AdditionalArguments = nil
+	mc.AddArguments("quarkus-bootstrap:build-tree")
```
Comment (lburgazzoli) [inline pkg/builder/runtime/quarkus.go]: do we need that ? as far as I know it should be enough to add runtime dependencies so it should be enough to use the standard dependency plugin

Comment (astefanutti) [reply pkg/builder/runtime/quarkus.go]: Ah my understanding was that the augmentation process could add runtime dependencies as well. Let me try using the standard dependency plugin then!

File: pkg/trait/environment.go [hunk 2]
```diff
@@ -63,7 +64,12 @@ func (t *environmentTrait) Apply(e *Environment) error {
 		envvar.SetValFrom(&e.EnvVars, envVarPodName, "metadata.name")
 	}
 
-	envvar.SetVal(&e.EnvVars, envVarMainClass, defaultMainClass)
+	quarkus := e.Catalog.GetTrait("quarkus").(*quarkusTrait)
+	if quarkus.isEnabled() {
+		quarkus.addContainerEnvironment(e)
+	} else {
+		envvar.SetVal(&e.EnvVars, envVarMainClass, defaultMainClass)
```
[the remark this answers is not in the document]
Comment (astefanutti) [reply pkg/trait/environment.go]: Absolutely. It's all scattered so I should be done properly.

File: docs/modules/ROOT/pages/traits.adoc [hunk 3]
```diff
@@ -394,6 +394,24 @@ It's disabled by default.
...
+It's disabled by default.
+
+[cols="m,"]
+!===
+
+! quarkus.camel-quarkus-version
+! The Camel Quarkus version to use for the integration
+
+! quarkus.quarkus-version
+! The Quarkus version to use for the integration
```
[the remark this answers is not in the document]
Comment (astefanutti) [reply docs/modules/ROOT/pages/traits.adoc]: Ultimately, the relation is captured in the matching catalog and taken from there. These parameters are only used to select the match the Camel catalog to be used and that is prescriptive in term of compatibility.

File: pkg/builder/runtime/quarkus.go (hunk 1 above)
Comment (lburgazzoli) [reply pkg/builder/runtime/quarkus.go]: Theoretically everything should be derived from the runtime artefacts, but need to be proven :)

File: deploy/camel-catalog-quarkus-0.3.0-1.0.7-SNAPSHOT.yaml [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+apiVersion: camel.apache.org/v1alpha1
+kind: CamelCatalog
+metadata:
+  name: camel-catalog-quarkus-0.3.0-1.0.7-snapshot
+  labels:
+    app: camel-k
+    camel.apache.org/catalog.version: 3.0.0-RC3
+    camel.apache.org/catalog.loader.version: 3.0.0-RC3
+    camel.apache.org/runtime.version: 1.0.7-SNAPSHOT
+    camel.apache.org/runtime.provider: quarkus
```
[the remark this answers is not in the document]
Comment (astefanutti) [reply deploy/camel-catalog-quarkus-0.3.0-1.0.7-SNAPSHOT.yaml]: The Quarkus versions are included in the `.spec.RuntimeProvider.Quarkus` struct, but we can add extra annotations for them.

[the remark this answers is not in the document]
Comment (lburgazzoli) [reply deploy/camel-catalog-quarkus-0.3.0-1.0.7-SNAPSHOT.yaml]: yep, it is just for consistency

File: pkg/builder/runtime/quarkus.go (hunk 1 above)
Comment (astefanutti) [reply pkg/builder/runtime/quarkus.go]: I've just done a quick test and that seems to work. So the theory seems valid at least on simple examples :)

File: deploy/camel-catalog-quarkus-0.3.0-1.0.7-SNAPSHOT.yaml [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+  version: 3.0.0-RC3
+  runtimeVersion: 1.0.7-SNAPSHOT
+  runtimeProvider:
+    quarkus:
+      camelQuarkusVersion: 0.3.0
+      quarkusVersion: 0.26.1
+  artifacts:
+    camel-k-loader-groovy:
+      groupId: org.apache.camel.k
+      artifactId: camel-k-loader-groovy
```
[the remark this answers is not in the document]
Comment (astefanutti) [reply deploy/camel-catalog-quarkus-0.3.0-1.0.7-SNAPSHOT.yaml]: My understanding is that it's covered in the `addRuntimeDependencies` of the quarkus trait.

**Pull request merged**